docs/docs/guides/05-type-system/02-concepts/03-public-registry.mdx

```diff
@@ -0,0 +1,316 @@
+}
+```
+
+Every resource comes with its own registry that has a set of rules acting as the default state (the index).
+
+Because every registry can be augmented (publicly exported), developers have the freedom to inject their own definitions.
+
+Doing so allows Strapi APIs to respond in a way that aligns with the developers' custom definitions.
+
+:::info[Did you know?]
+The `Schema` and `UID` namespaces rely solely on public registries to infer their type definitions.
+
+This is why it's heavily encouraged to use them over their low level counter-parts.
+:::
+
+### How it works
+
+#### Architecture
+
+```mermaid
+flowchart TB;
+  %% Main node
+  ContentTypeRegistry("Content-Types Registry")
+
+  %% App
+  subgraph Application
+    %% Nodes
+    UserApp["User Application"]
+    UserSchema(["Application Types"])
+
+    %% Links
+    UserApp -- "generates" ---> UserSchema
+  end
+
+  %% Type System
+  subgraph TypesPackage["@strapi/types"]
+    %% Nodes
+    TypeSystem["Type System"]
+    Types{{Types}}
+    Registries{{Registries}}
+    UID{{UID}}
+    Schema{{Schema}}
+
+    %% Links
+    TypeSystem -- "exports" ---> Types & Registries
+    Types -- "exports" ---> UID & Schema -- "uses" ---> ContentTypeRegistry
+    Registries -- "exports" ----> ContentTypeRegistry
+  end
+
+  %% Strapi
+  subgraph StrapiPackage["@strapi/strapi"]
+    %% Nodes
+    Strapi["Strapi"]
+    APIs[["APIs"]]
+
+    %% Links
+    Strapi -- "exports" --> APIs
+    APIs -- "uses" --> Types
+  end
+
+  %% This link needs to be placed last to preserve the layout
+  UserSchema -- "augments" -----> ContentTypeRegistry
+```
+
+#### Usage
+
+##### 1. Registry Definition
+
+Creating a new registry is as simple as exporting an indexed interface from the right namespace.
+
+Let's declare the content-type schema registry.
+
+It should accept:
+- Content-type UIDs as keys
+- Content-type schemas as values.
+
+```ts title="@strapi/types/public/registries.ts"
+import type { Internal, Struct } from '@strapi/types';
+
+export interface ContentTypeSchemas {
+  [key: Internal.UID.ContentType]: Struct.ContentTypeSchema;
+}
+```
+
+:::note
+We use low level types to define our index (`Internal`/`Struct`) to keep it as generic as possible.
+:::
+
+##### 2. Dynamic Type Definitions
+
+###### UID.ContentType
+
+To define `UID.ContentType`, we extract every key (`Internal.Registry.Keys`) from the public content-type registry (`Public.ContentTypeSchemas`) that match with the base definition of a content-type UID (`Internal.UID.ContentType`).
+
+```ts title="@strapi/types/uid.ts"
+import type { Internal, Public } from '@strapi/types'
+
+export type ContentType = Internal.Registry.Keys<
+  Public.ContentTypeSchemas,
+  Internal.UID.ContentType
+>;
+```
+
+:::note
+Only picking-up keys that extend `Internal.UID.ContentType` ensure we don't end up with manually added malformed keys, and tells the type-checker we're confident about what's in our union type.
```

Comment:
I'm not sure I understand this sentence - picking-up keys?

Reply:
We start with an object that has `key`/`value` pairs, then we hand-pick certain keys based on some conditions, and extract them into another type (a list of all those keys)

Reply:
```suggestion
Only selecting keys that extend `Internal.UID.ContentType` ensures we don't end up with manually added malformed keys, and tells the type-checker we're confident about what's in our union type.
```